### PR TITLE
M1504: Implement final charset support for text_response method

### PR DESCRIPTION
**Pull request by jitendra29**
Following changes have been made:
- Added final_mime_type and final_charset helper methods to return the appropriate final mime/charset values
- Modified the text_response method to use the final_charset instead of directly using the response charset

[Review on Reviewable](https://reviewable.io/reviews/servo/servo/8739)

### CODE REVIEW

File: components/script/dom/xmlhttprequest.rs [hunk 1]
```diff
@@ -995,21 +995,14 @@ impl XMLHttpRequest {
...
-                    if name == &mime::Attr::Charset {
-                        encoding = encoding_from_whatwg_label(&value.to_string()).unwrap_or(encoding);
-                    }
-                }
-            },
-            None => {}
+        let final_charset = self.final_charset();
+        if final_charset.is_some() {
+            encoding = final_charset.unwrap();
         }
```
Comment (eefriedman) [inline components/script/dom/xmlhttprequest.rs]: `let mut encoding = self.final_charset().unwrap_or(UTF_8)`?

**Comment by eefriedman** [comment]
```
/home/travis/build/servo/servo/components/script/dom/xmlhttprequest.rs:1115:5: 1126:6 warning: method is never used: `final_mime_type`, #[warn(dead_code)] on by default

/home/travis/build/servo/servo/components/script/dom/xmlhttprequest.rs:1115     fn final_mime_type(&self) -> Option<Mime> {
```

I guess you could just mark it with `#[allow(dead_code)]` for now...

File: components/script/dom/xmlhttprequest.rs [hunk 2]
```diff
@@ -1102,6 +1095,35 @@ impl XMLHttpRequest {
...
         Ok(())
     }
+
+    fn final_charset(&self) -> Option<EncodingRef> {
+        if self.override_charset.borrow().is_some() {
+            return self.override_charset.borrow().clone();
```
Comment (eefriedman) [inline components/script/dom/xmlhttprequest.rs]: You don't need to use "return" here; same in other places.

File: components/script/dom/xmlhttprequest.rs [hunk 3]
```diff
@@ -1102,6 +1095,35 @@ impl XMLHttpRequest {
...
+        } else {
+            match self.response_headers.borrow().get() {
+                Some(&ContentType(ref mime)) => {
+                    let value =  mime.get_param(mime::Attr::Charset);
```
Comment (eefriedman) [inline components/script/dom/xmlhttprequest.rs]: Whitespace

File: tests/wpt/mozilla/meta/css/iframe/size_attributes_vertical_writing_mode.html.ini [hunk 4]
```diff
@@ -1,3 +1,6 @@
 prefs: [layout.writing-mode.enabled:true]
 [viewport_meta.html]
   type: reftest
+
+[size_attributes_vertical_writing_mode.html]
+  type: reftest
```
Comment (eefriedman) [inline tests/wpt/mozilla/meta/css/iframe/size_attributes_vertical_writing_mode.html.ini]: What happened here?

Comment (dzbarsky) [inline tests/wpt/mozilla/meta/css/iframe/size_attributes_vertical_writing_mode.html.ini]: update-wpt put that there. Guess it's missing from the manifest or something.

Comment (Ms2ger) [inline tests/wpt/mozilla/meta/css/iframe/size_attributes_vertical_writing_mode.html.ini]: Typo on my side. Fixed in #8736. Please revert this change?

Comment (jitendra29) [inline tests/wpt/mozilla/meta/css/iframe/size_attributes_vertical_writing_mode.html.ini]: Yes update-wpt put it there, will revert back this change.

**Comment by jitendra29** [comment]
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[components/script/dom/xmlhttprequest.rs, line 1105 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8739#-K4KckGmoGpQJAfhH-aV-r1-1105:-K4PcoJRBWKCMjI4Ddtd:-1540910157) ([raw file](https://github.com/servo/servo/blob/3161a0aab0ce77f6e4c29d54a180d5c879091890/components/script/dom/xmlhttprequest.rs#L1105)):*
sorry didn't get where exactly. Can you point out the line number ?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8739)*

**Comment by eefriedman** [comment]
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[components/script/dom/xmlhttprequest.rs, line 1105 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8739#-K4KckGmoGpQJAfhH-aV-r1-1105:-K4PdZZkarC-7HI4ddFr:919637735) ([raw file](https://github.com/servo/servo/blob/3161a0aab0ce77f6e4c29d54a180d5c879091890/components/script/dom/xmlhttprequest.rs#L1105)):*
Extra space after equal sign in `value =  mime`...

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8739)*

**Comment by jitendra29** [comment]
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[components/script/dom/xmlhttprequest.rs, line 1004 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8739#-K4KckGmoGpQJAfhH-aV-r1-1004:-K4PkKd-kClpN3VxKhvX:-1100736181) ([raw file](https://github.com/servo/servo/blob/3161a0aab0ce77f6e4c29d54a180d5c879091890/components/script/dom/xmlhttprequest.rs#L1004)):*
we don't need encoding to be mutable if we use this right?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8739)*

**Comment by eefriedman** [comment]
*[components/script/dom/xmlhttprequest.rs, line 1004 \[r1\]](https://reviewable.io:443/reviews/servo/servo/8739#-K4KckGmoGpQJAfhH-aV-r1-1004:-K4Pr1Cjf4ByMxwtBuEo:641798574) ([raw file](https://github.com/servo/servo/blob/3161a0aab0ce77f6e4c29d54a180d5c879091890/components/script/dom/xmlhttprequest.rs#L1004)):*
Oh, yes, you're right.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/8739)*

**Comment by jitendra29** [comment]
I pulled the latest changes from servo/master and ran the tests without any code changes. I am getting the following unexpected test results :

<code>
./mach test-wpt XMLHttpRequest/ --release
Running 186 tests in web-platform-tests

  [73/186] /XMLHttpRequest/overridemimetype-headers-received-state-force-shiftji 
 ▶ TIMEOUT [expected OK] /XMLHttpRequest/event-upload-progress-crossorigin.sub.htm

  ▶ Unexpected subtest result in /XMLHttpRequest/event-upload-progress-crossorigin.sub.htm:
  │ TIMEOUT [expected PASS] XMLHttpRequest: upload progress event for cross-origin requests
  └   → Test timed out

  [103/186] /XMLHttpRequest/overridemimetype-open-state-force-xml.htm
  [104/186] /XMLHttpRequest/overridemimetype-open-state-force-xml.htmr-existing-  [104/186] /XMLHttpRequest/overridemimetype-open-state-force-xml.htmr-existing-  [105/186] /XMLHttpRequest/overridemimetype-open-state-force-xml.htmr-existing-  [105/186] /XMLHttpRequest/overridemimetype-open-state-force-xml.htmr-existing-  [106/186] /XMLHttpRequest/overridemimetype-open-state-force-xml.htmr-existing-  [106/186] /XMLHttpRequest/overridemimetype-open-state-force-xml.htmr-existing-  

▶ Unexpected subtest result in /XMLHttpRequest/send-redirect-bogus-sync.htm:g-  │ FAIL [expected PASS] XMLHttpRequest: send() - Redirects (bogus Location header; sync) (302: http://z)
  │   → assert_throws: function "function () { client.send(null) }" did not throw
  │ 
  │ redirect/<@http://web-platform.test:8000/XMLHttpRequest/send-redirect-bogus-sync.htm:5:11
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1381:20
  │ test@http://web-platform.test:8000/resources/testharness.js:495:9
  │ redirect@http://web-platform.test:8000/XMLHttpRequest/send-redirect-bogus-sync.htm:2:1
  └ @http://web-platform.test:8000/XMLHttpRequest/send-redirect-bogus-sync.htm:9:7

  ▶ Unexpected subtest result in /XMLHttpRequest/send-redirect-bogus.htm:
  │ FAIL [expected PASS] XMLHttpRequest: send() - Redirects (bogus Location header) (302: http://example.not)
  │   → assert_equals: expected 0 but got 200
  │ FAIL [expected PASS] XMLHttpRequest: send() - Redirects (bogus Location header) (303: http://example.not)
  │   → assert_equals: expected 0 but got 200
  │ 
  │ redirect/</client.onreadystatechange/<@http://web-platform.test:8000/XMLHttpRequest/send-redirect-bogus.htm:8:17
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1381:20
  └ redirect/</client.onreadystatechange@http://web-platform.test:8000/XMLHttpRequest/send-redirect-bogus.htm:6:13

Ran 186 tests finished in 198.0 seconds.st-timeout-worker-synconworker.html
  • 183 ran as expected. 4 tests skipped.
  • 1 tests timed out unexpectedly
  • 3 tests had unexpected subtest results
</code>

I shouldn't ideally get these because I didn't make any code review changes yet and all tests were passing when I made the pull request.

**Comment by jdm** [comment]
Did you build with `./mach build --release` before running that?

**Comment by jitendra29** [comment]
@jdm yes I did !

**Comment by jitendra29** [comment]
@eefriedman  We have made the suggested changes to the pull request
